Account for shard offset within the `run` subcommand

manuelmauro opened the pull request:
## Description

Moonwall's `run` subcommand stopped working due to port mismatch introduced in https://github.com/Moonsong-Labs/moonwall/pull/498.
This PR fixes the issue by accounting for shard offset within the `run` subcommand.

- Use singleton class instead of environment variable to manage sharding.
- Adds port information to error logs to allow checking for port mismatches.

timbrinded commented:
just a reminder that running moonwall `run` command often breaks for big runs since it doesn't support multi-sharding and it doesn't refresh state between runs. It's only really designed for testing individual tests

timbrinded commented:
@codex do a review

timbrinded commented:
@codex review this!

timbrinded commented:
Depending on what you are trying to do here, test shards might be irrelevant. Test sharding takes test files: [ a.ts, b.ts, c.ts] and turns it into two subsets of runs which is useful for the CI to pre-split which tests to run: 
- RunnerA: [a.ts, b.ts]
- RunnerB: [c.ts]

---

Port conflicts happen you have multiple nodes running on the same port internally. Previously i used environment variable `VITEST_POOL_ID`, which is what vitest adds to the launched child process envVars when it spawns it. That way you can deterministically make sure you don't pick the same port without having to communicate between threads. 

Node.js's event loop does not have shared memory between threads, so a singleton shardmanager might be hard to accomplish unless you use IPX to do interprocess communication and locks.

---

`run` should be enforced single threaded because `MOON_RECYCLE` env var forces it:

https://github.com/Moonsong-Labs/moonwall/blob/c9defca4693938af7517461f03e317c87e653a7a/packages/cli/src/lib/globalContext.ts#L422

manuelmauro commented:
This PR is just a fix to a regression introduced in https://github.com/Moonsong-Labs/moonwall/pull/498.
There we improved the port offset calculation:

```typescript
// Calculate port with better isolation between shards
// Base port 10000 + (shard * 1000) + (pool * 100) + deterministic offset
```

and introduced port scanning:

```typescript
return getNextAvailablePort(startPort);
```

Due to those changes the run subcommand stopped working because `process.env.MOONWALL_TEST_SHARD` was undefined within `async function runNetworkCmd(args: RunCommandArgs)`.

Here I introduced a singleton because it seems like a cleaner way to define a global variable instead of using `process.env`, but there might be better solutions and there might even be a more idiomatic way to define globals within Moonwall that I missed.

We use this global variable in `export const getFreePort` instead of passing arguments down the call stack. We do not use this information across threads and do not rely on it in order to avoid port collisions.

timbrinded reviewed (approved):
LGTM
